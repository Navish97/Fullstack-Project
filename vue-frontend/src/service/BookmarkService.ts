import axiosInstance from "@/service/AxiosInstance";

export async function deleteBookmark(itemId: number): Promise<any> {
    return await axiosInstance.delete(`/api/bookmarks/delete/${itemId}`);
}

export async function addBookmark(itemId: number): Promise<any> {
    return await axiosInstance.post(`/api/bookmarks/add/${itemId}`);
}

export async function getUserBookmarks(): Promise<any> {
    return await axiosInstance.get(`/api/bookmarks/user`);
}