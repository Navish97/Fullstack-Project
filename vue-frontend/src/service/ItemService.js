import axios from "axios";

const itemAPI = axios.create({
    baseURL: 'http://localhost:8080/api/items',
    withCredentials:true,
    headers: {
        Accept:'application/json',
        'Content-Type':'application/json',
    }
})

export const getItems = (pageNr, pageSize) => {
        return itemAPI.get('/page', {
            params:{
                pageNumber:pageNr,
                size:pageSize,
            }
        });
}

