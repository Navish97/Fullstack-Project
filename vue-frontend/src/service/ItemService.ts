import axiosInstance from "@/service/AxiosInstance";

export async function getItems(pageNr : number, pageSize : number, filter : object): Promise<any> {
    const response = await axiosInstance.get('/api/items/page', {
        params: {
            pageNumber:pageNr,
            size:pageSize,
            filter:JSON.stringify(filter)
        }
    }).then((response) => {
        return response;
    })
    .catch((error) => {
        console.log('Error loading items: ', error.message)
    })
    return response;
}

export async function getItemDetails(itemId: String): Promise<any> {
    return await axiosInstance.get(`/api/items/details/${itemId}`);
}