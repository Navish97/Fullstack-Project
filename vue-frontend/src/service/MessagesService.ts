import axiosInstance from "./AxiosInstance";
import type { Chat } from '@/types/ChatType';

export async function getChats():Promise<any> {
    const response = await axiosInstance.get('/api/chats/load-chats')
    .then((response) => {
        return response;
    })
    .catch((error) => {
        console.log('Error retrieving chats', error.message);
    })
    return response;
}

export async function getMessages(chat : Chat):Promise<any> {
    const response = await axiosInstance.get('/api/chats/load-messages', {
        params: {
            chatId : chat.id
        },
    })
    .then((response) => {
        return response;
    })
    .catch((error) => {
        console.log('Error retrieving chats', error.message);
    })
    return response;
}

export async function sendMessage(message: String, chatId:Number, itemId:Number):Promise<any> {
    const response = axiosInstance.post('/api/chats/send-message', 
    {
        message:message,
        chatId:chatId,
        itemId:itemId,
        })
    return response;
}