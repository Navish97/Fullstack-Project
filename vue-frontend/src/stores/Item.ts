import { defineStore } from 'pinia'
import type { Item } from "@/types/ItemType";

export const useItemStore = defineStore({
  id: 'items',
  state: () => ({
    listingTypes: ['thumbnail', 'list'] as string[],
    currentListingType: 'thumbnail' as string,
    currentItem: {} as Item,
    persist: {
      storage: sessionStorage,
    },
    items : [
   ] as Item[],
    NewListingCategory: 0 as number,
  }),
  getters: {
    getCurrentItem: (state) => {
      return state.currentItem;
    },
    getCurrentItemId: (state) => {
        return state.currentItem.id;
    },
    getCurrentListingType: (state) => {
      return state.currentListingType;
    },
    getListingTypes: (state) => {
      return state.listingTypes;
    },
    getItems: (state) => {
      return state.items;
    },
    getNewListingCategory: (state) => {
      return state.NewListingCategory;
    }
  },
  actions: {
    setCurrentItem(item: Item) {
      this.currentItem = item;
    },
    setCurrentListingType(listingType: string) {
      this.currentListingType = listingType;
    },
    addItem(item : Item){
      const currentItems = this.items;
      const updatedItems = [...currentItems, item];
      this.items = updatedItems;
    },
    setLists(list : []){
      const newItems:Item[] = [];
      list.forEach((element) => {
        const data = JSON.parse(JSON.stringify(element));
        const newItem = {
          id: data.id,
          description: data.description,
          briefDescription: data.briefDescription,
          price: data.price,
          imageURLs: [data.imageUrls],
          categoryId: data.categoryid,
          title: data.title,
          latitude: parseFloat(data.latitude),
          longitude: parseFloat(data.longitude),
          userId: data.userid,
        }
        newItems.push(newItem);
      });
      this.items = newItems;
    },
    setNewListingCategory(categoryId: number){
      this.NewListingCategory = categoryId;
    }
  },
});
