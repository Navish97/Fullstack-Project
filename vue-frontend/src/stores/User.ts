import { defineStore } from 'pinia'
import type { Bookmark } from "@/types/BookmarkType";
import type {Item} from "@/types/ItemType";
import router from "@/router";

export const useUserStore = defineStore({
    id: 'user',
    state: () => ({
        loggedInUserEmail: "" as string,
        bookmarks: [] as Bookmark[],
    }),
    persist: {
        storage: sessionStorage,
    },
    getters: {
        getLoggedInUserEmail: (state) => {
            return state.loggedInUserEmail;
        },
        getBookmarks: (state) => {
            return state.bookmarks
        },
        isItemBookmarked: (state) => (item: Item) => {
                return state.bookmarks.some(bookmark => bookmark.itemId === item.id);
        },
        addBookmark: (state) => (item: Item) => {
            //Axios call for adding bookmark to database
        },
        isLoggedIn: (state) => () => {
            return state.loggedInUserEmail !== "";
        }

    },
    actions: {
        logOut() {
            router.push("/")
            this.loggedInUserEmail = "";
            this.bookmarks = [];
        },
        setLoggedInUserEmail(userEmail: string) {
            this.loggedInUserEmail = userEmail;
        },
        setBookmarks(bookmarks: Bookmark[]) {
            this.bookmarks = bookmarks;
        },
    }
})