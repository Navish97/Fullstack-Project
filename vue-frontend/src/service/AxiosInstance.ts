import axios from "axios";
import type { AxiosInstance } from "axios";
import { useUserStore } from "@/stores/User";
const baseURL = window.location.hostname === "localhost" ? "http://localhost:8080" : "https://fullstack-project-xt5ws57jza-lz.a.run.app";

const axiosInstance: AxiosInstance = axios.create({
    baseURL,
    withCredentials:true,
    timeout: 5000,
    headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
    }
});

axiosInstance.interceptors.response.use(
    (response) => {
        return response;
    },
    (error) => {
        console.log(error.response.headers)
        if (error.response.headers['error-message']) {
            alert("Old password is incorrect. Please try again.");
            return;
        }
        else if (error.response.status === 401 && useUserStore().isLoggedIn) {
            console.log("Session has expired. You've been logged out.");
            alert("Session has expired. You've been logged out.");
            useUserStore().logOut()
        }
        else if(error.response.status !== 401) {
            console.log("An error has occurred. Errorcode: " + error.response.status);
        }
        return Promise.reject(error);
    }
);
export default axiosInstance;