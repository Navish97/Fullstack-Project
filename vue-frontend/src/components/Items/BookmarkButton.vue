<template>
  <div class="container" v-if="isLoggedIn" >
    <button class="bookmark-button" :class="{ 'bookmarked': isBookmarked }" @click="toggleBookmark">
      <span class="bookmark-icon"></span>
      <span class="bookmark-text">{{ isBookmarked ? 'Remove Bookmark' : 'Add Bookmark' }}</span>
    </button>
  </div>
</template>

<script setup lang="ts">
import {computed} from "vue";
import {useItemStore} from "@/stores/Item";
import {useUserStore} from "@/stores/User";
import {addBookmark, deleteBookmark} from "@/service/BookmarkService";

const itemStore = useItemStore();
const userStore = useUserStore();

async function toggleBookmark() {
  if (itemStore.isCurrentItemBookmarked) {
    const response = await deleteBookmark(itemStore.getCurrentItemId);
    if (response.status === 200) {
      itemStore.setCurrentItemBookmarked(false);
    }
  } else {
    const response = await addBookmark(itemStore.getCurrentItemId);
    if (response.status === 200) {
      itemStore.setCurrentItemBookmarked(true);
    }
  }
}

const isLoggedIn = computed(() => {
  return userStore.isLoggedIn();
});

const isBookmarked = computed(() => {
  return itemStore.isCurrentItemBookmarked;
});
</script>

<style scoped>
.container {
  padding-top: 8px;
}
.bookmark-button {
  display: inline-block;
  padding: 8px 8px;
  border-radius: 4px;
  border: 1px solid #000000;
  background-color: #1c1b1b;
  color: #FFFFFF;
  font-size: 16px;
  font-weight: bold;
  text-transform: uppercase;
  cursor: pointer;
  transition: background-color 0.3s ease;
}

.bookmark-button:hover {
  background: #4d4d4d;
}

.bookmark-button.bookmarked {
  color: #fafafa;
  background: #425432;
}

.bookmark-button.bookmarked:hover {
  background: #1a6e27;
}


.bookmark-text {
  display: inline-block;
  vertical-align: middle;
}
</style>