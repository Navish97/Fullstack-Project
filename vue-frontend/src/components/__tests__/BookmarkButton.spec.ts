import { beforeEach, expect, it, vi, afterEach } from "vitest";
import { createPinia } from "pinia";
import { mount } from "@vue/test-utils";
import BookmarkButton from "@/components/Items/BookmarkButton.vue";
import { useItemStore } from "@/stores/Item";
import { useUserStore } from "@/stores/User";
import * as BookmarkService from "@/service/BookmarkService";

let wrapper: any;
let addBookmarkSpy: any;
let deleteBookmarkSpy: any;

beforeEach(() => {
    addBookmarkSpy = vi.spyOn(BookmarkService, "addBookmark").mockImplementation(() => {
        return Promise.resolve({ status: 200 });
    });

    deleteBookmarkSpy = vi.spyOn(BookmarkService, "deleteBookmark").mockImplementation(() => {
        return Promise.resolve({ status: 200 });
    });

    const pinia = createPinia();

    const itemStore = useItemStore(pinia);
    itemStore.setCurrentItemBookmarked(false);

    const userStore = useUserStore(pinia);
    userStore.setLoggedIn(true);

    wrapper = mount(BookmarkButton, { global: { plugins: [pinia] } });
});




it("It renders the button properly", () => {
    expect(wrapper.html()).toMatchSnapshot();
});


it("It toggles the bookmark state on button click", async () => {
    const itemStore = useItemStore(wrapper.vm.$pinia);
    itemStore.setCurrentItemBookmarked(false);

    await wrapper.find(".bookmark-button").trigger("click");
    expect(itemStore.isCurrentItemBookmarked).toBe(true);

    await wrapper.find(".bookmark-button").trigger("click");
    expect(itemStore.isCurrentItemBookmarked).toBe(false);
});

it("It displays the correct text based on bookmark state", () => {
    const itemStore = useItemStore(wrapper.vm.$pinia);

    itemStore.setCurrentItemBookmarked(false);
    wrapper = mount(BookmarkButton, { global: { plugins: [wrapper.vm.$pinia] } });

    expect(wrapper.text()).toContain("Add Bookmark");

    itemStore.setCurrentItemBookmarked(true);
    wrapper = mount(BookmarkButton, { global: { plugins: [wrapper.vm.$pinia] } });

    expect(wrapper.text()).toContain("Remove Bookmark");
});

it("It does not render the button if the user is not logged in", () => {
    const userStore = useUserStore(wrapper.vm.$pinia);
    userStore.setLoggedIn(false);

    wrapper = mount(BookmarkButton, { global: { plugins: [wrapper.vm.$pinia] } });

    expect(wrapper.find(".bookmark-button").exists()).toBe(false);
});
