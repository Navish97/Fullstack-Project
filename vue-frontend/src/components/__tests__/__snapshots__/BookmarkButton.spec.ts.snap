// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`It renders the button properly 1`] = `"<div data-v-769e16e1=\\"\\" class=\\"container\\"><button data-v-769e16e1=\\"\\" class=\\"bookmark-button\\"><span data-v-769e16e1=\\"\\" class=\\"bookmark-icon\\"></span><span data-v-769e16e1=\\"\\" class=\\"bookmark-text\\">Add Bookmark</span></button></div>"`;
